Enable Antrea in GKE cluster for Ubuntu host (#532)
vmware-tanzu/antrea: docs/gke-installation.md

# Deploying Antrea on a GKE cluster

We support running Antrea inside of GKE clusters on Ubuntu host. Antrea would operate

> in noEncap mode, in which no encapsulation is required for any kind of traffic (Intra Node,

Reviewer: To be consistent with Su's implementation, call it NetworkPolicy only mode.

Author: Actually I want to highlight noEncap mode like Calico does in its documentation too. Can we keep this as is.  Also, would @antoninbas be updating getting started guide, or I should do it?

Reviewer 2: you should upload all the relevant documentation as part of this PR

Reviewer: noEncap mode is confusing as in our implementation it means different thing. I think we should use policy only mode, but you can explain about no encapsulation.

Author: Done. Changed to NetworkPolicy only mode.

> Inter Node etc) and enforcing network policies using OVS.

Reviewer: Need to revise the last sentence, like: and NetworkPolicies are enforced using OVS

Author: Done

## GKE Prerequisites

1. Install the Google Cloud SDK (gcloud). Refer to this (https://cloud.google.com/sdk/install)

    ```bash
    curl https://sdk.cloud.google.com | bash
    ```

2. Make sure you are authenticated to use the Google Cloud API

    ```bash
    export ADMIN_USER=user@email.com
    gcloud auth login
    ```

3. Create a project or use an existing one

    ```bash
    export GKE_PROJECT=gke-clusters
    gcloud projects create $GKE_PROJECT
    ```

## Creating the cluster

You can use any method to create a GKE cluster (gcloud SDK, gcloud Console etc). The example
given here is using the Google Cloud SDK.

> **Note:** Antrea is supported on Ubuntu hosts only for GKE cluster. Also, its must to select service

Reviewer: hosts to Nodes?

Reviewer: it is a must?

Author: done

    export GKE_HOST="UBUNTU"
    export GKE_SERVICE_CIDR="10.94.0.0/16"
    gcloud container --project $GKE_PROJECT clusters create cluster1 --image-type $GKE_HOST \
       --zone $GKE_ZONE --enable-ip-alias --services-ipv4-cidr $GKE_SERVICE_CIDR
    ```

2. Access your cluster

    ```bash
    kubectl get nodes
    NAME                                      STATUS   ROLES    AGE     VERSION
    gke-cluster1-default-pool-93d7da1c-61z4   Ready    <none>   3m11s   v1.14.10-gke.17
    gke-cluster1-default-pool-93d7da1c-rkbm   Ready    <none>   3m9s    v1.14.10-gke.17
    ```

3. Create a cluster-admin ClusterRoleBinding

    ```bash
    kubectl create clusterrolebinding cluster-admin-binding --clusterrole cluster-admin --user user@email.com
    ```

    **Note:** To create clusterRoleBinding, the user must have `container.clusterRoleBindings.create` permission.
Use this command to enable it, if the previous command fails due to permission error. Only cluster Admin can
assign this permission.

    ```bash
    gcloud projects add-iam-policy-binding $GKE_PROJECT --member user:user@email.com --role roles/container.admin
    ```

## Deploying Antrea

1. Prepare the Cluster Nodes

>     Deploy antrea-node-init DaemonSet to enable kubelet to operate in CNI mode.

Reviewer: Mark specific names with `` or ````, like `antrea-node-init`, `kubelet`. Please check other documents.

Author: Done

    ```bash

>     kubectl apply -f https://github.com/vmware-tanzu/antrea/releases/download/v0.5.0/patches/gke/antrea-node-init.yml

Author: @antoninbas @tnqn @abhiraut Will this be a valid link ??? Please check.

Reviewer 3: @antoninbas , Is there any possible we have a latest tag of release? Thus people could always get newest release from the same URL, like ` https://github.com/vmware-tanzu/antrea/releases/download/latest/patches/gke/antrea-node-init.yml`.

Reviewer: Please check ipsec-tunnel.md and eks-installation.md to describe yaml files.

Reviewer 2: Unfortunately, there is no link "alias" for the latest published release.

@reachjainrahul this link will definitely *not* work. There is no directory structure for assets, so *if* we wanted to have this as an asset, the link would have to be something like `https://github.com/vmware-tanzu/antrea/releases/download/v0.5.0/antrea-gke-node-init.yml`. However, there is no reason to publish this as an asset here: the reason why we publish Antrea yaml manifests as assets is because the release version is different from the dev version (e.g. we use a different Antrea docker image tag). In the case of antrea-node-init.yml, there is a single version of the yaml (it's pretty much never going to change), so you should use a "regular github tree link": based on the current location, it will be `https://github.com/vmware-tanzu/antrea/tree/v0.5.0/build/yamls/patches/gke/antrea-node-init.yml`. However, I don't think that placing it under "patches" is appropriate here.

Reviewer: Yes we will publish the yaml from 0.5.0. But here I suggest you to follow description of other yaml files on how to get the yaml - link for the latest one, and that for a released version. You might consider some reorganization of your descriptions to talk about released versions only once for antrea-gke-node-init.yml and antrea-gke.yml.

Author: I am keeping antrea-gke-node-init documentation as is and updated antrea-gke.yml documentation to reflect latest and released version. @antoninbas I hope this link would work.  https://github.com/vmware-tanzu/antrea/releases/download/v0.5.0/antrea-gke-node-init.yml

    ```

2. Download and Update Antrea YAML

    Download a released version of Antrea deployment YAML (>=v0.5.0). Update the `serviceCIDR` config in the
antrea-gke.yml file with GKE_SERVICE_CIDR selected at the time of deploying GKE cluster.

    ```bash
    export TAG=v0.5.0
    curl https://github.com/vmware-tanzu/antrea/releases/download/<TAG>/antrea-gke.yml --output antrea-gke.yml

    cat antrea-gke.yml
    ...
        # CIDR Range for services in cluster. It's required to support egress network policy, should
        # be set to the same value as the one specified by --service-cluster-ip-range for kube-apiserver.
        serviceCIDR: 10.94.0.0/16
    ...
    ```

3. Deploy Antrea

>     Deploy Antrea using `kubectl apply -f antrea-gke.yml`. It will deploy single replica of Antrea controller

Reviewer: a single replica of

Author: Done

> and Antrea agent to every worker nodes of the GKE cluster. After a successful deployment you should be able to see

Reviewer: To be clear, maybe: and deploy Antrea agent to...

Reviewer: nodes -> Nodes

Author: > To be clear, maybe: and deploy Antrea agent to...

Didnt get you?

Reviewer: I meant this sentence is not very clear to me:
> It will deploy a single replica of Antrea controller and Antrea agent to every worker nodes of the GKE cluster.
Might be better:
> It will deploy a single replica of Antrea controller to the GKE cluster and deploy Antrea agent to every Node.

Author: Done

these Pods running in your cluster:

    ```bash

>     $ kubectl get pods --namespace kube-system  -l app=antrea -o wide

Reviewer: Suggest to refer to getting-started.md (there is a gif there), instead of repeating the same thing again.

Author: That Gif describe the installation step. I feel its OK to list it here as is. We anyway have to update this with final GKE proposal right?

# Deploying Antrea on a GKE cluster

> We support running Antrea inside of GKE clusters on Ubuntu host. Antrea would operate

Reviewer: host -> Node?

Author: Done

in noEncap mode, in which no encapsulation is required for any kind of traffic (Intra Node,

> Inter Node etc) and NetworkPolicies are enforced using OVS.

Reviewer: Inter Node, etc.)

Author: Done

## GKE Prerequisites

> 1. Install the Google Cloud SDK (gcloud). Refer to this (https://cloud.google.com/sdk/install)

Reviewer: Consider: Refer to [Google Cloud SDK installation guide](https://cloud.google.com/sdk/install).

Author: Done

    ```bash
    curl https://sdk.cloud.google.com | bash
    ```

2. Make sure you are authenticated to use the Google Cloud API

    ```bash
    export ADMIN_USER=user@email.com
    gcloud auth login
    ```

3. Create a project or use an existing one

    ```bash
    export GKE_PROJECT=gke-clusters
    gcloud projects create $GKE_PROJECT
    ```

## Creating the cluster

You can use any method to create a GKE cluster (gcloud SDK, gcloud Console etc). The example
given here is using the Google Cloud SDK.

> **Note:** Antrea is supported on Ubuntu Nodes only for GKE cluster. Also, it is must to select service

Reviewer: In my understanding, it should be "it is a must" but not "it is must", but I trust you will figure out the right grammar.

Author: Done

       --zone $GKE_ZONE --enable-ip-alias --services-ipv4-cidr $GKE_SERVICE_CIDR
    ```

2. Access your cluster

    ```bash
    kubectl get nodes
    NAME                                      STATUS   ROLES    AGE     VERSION
    gke-cluster1-default-pool-93d7da1c-61z4   Ready    <none>   3m11s   v1.14.10-gke.17
    gke-cluster1-default-pool-93d7da1c-rkbm   Ready    <none>   3m9s    v1.14.10-gke.17
    ```

3. Create a cluster-admin ClusterRoleBinding

    ```bash
    kubectl create clusterrolebinding cluster-admin-binding --clusterrole cluster-admin --user user@email.com
    ```

    **Note:** To create clusterRoleBinding, the user must have `container.clusterRoleBindings.create` permission.
Use this command to enable it, if the previous command fails due to permission error. Only cluster Admin can
assign this permission.

    ```bash
    gcloud projects add-iam-policy-binding $GKE_PROJECT --member user:user@email.com --role roles/container.admin
    ```

## Deploying Antrea

1. Prepare the Cluster Nodes

    Deploy ``antrea-node-init`` DaemonSet to enable ``kubelet`` to operate in CNI mode.

    ```bash

>     kubectl apply -f https://github.com/vmware-tanzu/antrea/releases/download/v0.5.0/antrea-gke-node-init.yml

Reviewer: Again, maybe we talk about the location of the yaml files of the latest and released versions before steps.

Like:
> Antrea is deployed using the deployment yamls. The latest version of the yaml files can be found in https://raw.githubusercontent.com/vmware-tanzu/antrea/master/build/yamls/. Deploy a released version of Antrea from the [list of releases](https://github.com/vmware-tanzu/antrea/releases). Note that GKE support was added in release 0.5.0, which means you can not pick a release older than 0.5.0.

Then in step 1 and 2:
> To deploy a given release `<TAG>` (e.g. `v0.5.0`):
> kubectl apply -f https://github.com/vmware-tanzu/antrea/releases/download/<TAG>/antrea-gke-node-init.yml

> To deploy the latest version:
> kubectl apply -f https://raw.githubusercontent.com/vmware-tanzu/antrea/master/build/yamls/antrea-gke-node-init.yml

Author: I prefer to keep antrea-gke-node-init.yml installation simple as is. We dont need to talk about release version for this, as we would be supporting just one version. If you prefer I can change the link to master, instead of 0.5.0.

Reviewer: Yes, changing to master is at least better.

Reviewer 2: I'm not planning to upload this to release assets, as I don't see a reason to. So the `releases/download` link will not be available. Since this file it not expected to change, +1 to Jianjun's suggestion.

Reviewer: @antoninbas : do you suggest to remove the released version for antrea-gke.yml too?

Reviewer 2: Yes I suggest just having `https://raw.githubusercontent.com/vmware-tanzu/antrea/master/build/yamls/antrea-gke-node-init.yml` in the instructions, as it's easier. If you really want a "released" link, you can use `https://github.com/vmware-tanzu/antrea/tree/<TAG>/build/yamls/antrea-gke-node-init.yml`. But I am not planning on making `https://github.com/vmware-tanzu/antrea/releases/download/<TAG>/antrea-gke-node-init.yml` a valid link.

Reviewer 2: This is based on the fact that this manifest does not point to any versioned Antrea resource (e.g. docker image) and is unlikely to change in ways that matter. If you think having a "release" link in the instructions is safer / more future-proof, feel free to use the link I paster above.

    ```

2. Download and Update Antrea YAML

    Deploy a released version of Antrea from the [list of releases](https://github.com/vmware-tanzu/antrea/releases).
Note that GKE support was added in release 0.5.0, which means you can not pick a release older than 0.5.0.
For any given release `<TAG>` (e.g. `v0.5.0`), get the Antrea GKE deployment yaml at:

    ````
    https://github.com/vmware-tanzu/antrea/releases/download/<TAG>/antrea-gke.yml
    ````

    To deploy the latest version of Antrea (built from the master branch) to GKE, get the Antrea GKE deployment yaml at:

    ````
    https://raw.githubusercontent.com/vmware-tanzu/antrea/master/build/yamls/antrea-gke.yml
    ````

>     Update ``serviceCIDR`` value of antrea-agent.conf in antrea-gke.yml with GKE_SERVICE_CIDR selected at the time of deploying GKE cluster.

Reviewer: Does GKE use different MTU like jumbo frame? If so, we should update the MTU parameter too?

Author: Updated it.

# Deploying Antrea on a GKE cluster

We support running Antrea inside of GKE clusters on Ubuntu Node. Antrea would operate
in NetworkPolicy only mode, in which no encapsulation is required for any kind of traffic
(Intra Node, Inter Node, etc) and NetworkPolicies are enforced using OVS. Antrea is supported
on both VPC-native Enable/Disable modes.

## GKE Prerequisites

1. Install the Google Cloud SDK (gcloud). Refer to [Google Cloud SDK installation guide](https://cloud.google.com/sdk/install)

    ```bash
    curl https://sdk.cloud.google.com | bash
    ```

2. Make sure you are authenticated to use the Google Cloud API

    ```bash
    export ADMIN_USER=user@email.com
    gcloud auth login
    ```

3. Create a project or use an existing one

    ```bash
    export GKE_PROJECT=gke-clusters
    gcloud projects create $GKE_PROJECT
    ```

## Creating the cluster

> You can use any method to create a GKE cluster (gcloud SDK, gcloud Console etc). The example

Reviewer 2: ```suggestion
You can use any method to create a GKE cluster (gcloud SDK, gcloud Console, etc). The example
```

    gke-cluster1-default-pool-93d7da1c-61z4   Ready    <none>   3m11s   v1.14.10-gke.17
    gke-cluster1-default-pool-93d7da1c-rkbm   Ready    <none>   3m9s    v1.14.10-gke.17
    ```

3. Create a cluster-admin ClusterRoleBinding

    ```bash
    kubectl create clusterrolebinding cluster-admin-binding --clusterrole cluster-admin --user user@email.com
    ```

    **Note:** To create clusterRoleBinding, the user must have `container.clusterRoleBindings.create` permission.
Use this command to enable it, if the previous command fails due to permission error. Only cluster Admin can
assign this permission.

    ```bash
    gcloud projects add-iam-policy-binding $GKE_PROJECT --member user:user@email.com --role roles/container.admin
    ```

## Deploying Antrea

1. Prepare the Cluster Nodes

    Deploy ``antrea-node-init`` DaemonSet to enable ``kubelet`` to operate in CNI mode.

    ```bash
    kubectl apply -f https://raw.githubusercontent.com/vmware-tanzu/antrea/master/build/yamls/antrea-gke-node-init.yml
    ```

2. Download and Update Antrea YAML

    Deploy a released version of Antrea from the [list of releases](https://github.com/vmware-tanzu/antrea/releases).

> Note that GKE support was added in release 0.5.0, which means you can not pick a release older than 0.5.0.

Reviewer 2: ```suggestion
Note that GKE support was added in release 0.5.0, which means you cannot pick a release older than 0.5.0.
```
